# use class to encapsulate scraping functions

2021/01/03

In [1]:
# run python scripts containing all the functions
%run -i '0_paths_open_save.py'
%run -i '1_scraping_cleaning_merge.py'
%run -i '2_summary_stats_visualization_new.py'

## 1. Scraping

In [2]:
initial_panels = ['201812', '201907', '201908', '201912', '202001', '202002']
current_panel = '202102'

for i in initial_panels:
    o = open_files(initial_panel=i)
    DF = o.open_df_to_id_for_scraping()
    id_list = get_id_from_data_before_202009(DF)
    print(i)
    print('The number of IDs are:', len(id_list))
    print('start scraping apps with initial panel', i)
    app_details = scraping_apps_according_to_id(id_list)
    print('Finished scraping app details with initial panel', i)
    s = save_files(initial_panel=i, current_panel=current_panel, app_details_dict=app_details)
    s.save_scraped_app_details()
    print('Saved scarped app details with initial panel', i)

scraping...: 100%|██████████| 8042/8042 [57:29<00:00,  2.33it/s]   


201812
The number of IDs are: 5977
start scraping apps with initial panel 201812
Finished scraping app details with initial panel 201812
Saved scarped app details with initial panel 201812
201907
The number of IDs are: 13144
start scraping apps with initial panel 201907
Finished scraping app details with initial panel 201907
Saved scarped app details with initial panel 201907
201908
The number of IDs are: 13015
start scraping apps with initial panel 201908
Finished scraping app details with initial panel 201908
Saved scarped app details with initial panel 201908
201912
The number of IDs are: 8468
start scraping apps with initial panel 201912
Finished scraping app details with initial panel 201912
Saved scarped app details with initial panel 201912
202001
The number of IDs are: 8359
start scraping apps with initial panel 202001
Finished scraping app details with initial panel 202001
Saved scarped app details with initial panel 202001
202002
The number of IDs are: 8042
start scraping app

## 2. Check the scraped data

In [2]:
current_panel = '201912'
tracking_panels = ['201812', '201907', '201908']

o = open_files(current_panel=current_panel, tracking_panels=tracking_panels)
app_details = o.open_app_detail_dict()
A = app_detail_dicts(app_details)
A.get_a_glimpse()

PANEL 201812
a random app id is: com.webruli.edmdjgame
title  :  Dj EDM Pads Game
icon  :  https://lh3.googleusercontent.com/DnDVnhzJcI0RfbtkBRVpcv6CXpqGbKP42IDgIgrac2VlOfQCBNu4J6wKLqSA3gfFBBs
screenshots  :  ['https://lh3.googleusercontent.com/wXvJVZPUjvCZ0NDqcfMwSbfTyl-CAVyokoZD8l5wij8PdQ8RRHiUuLca3KaPSMKK60w=w720-h310-rw', 'https://lh3.googleusercontent.com/EKEWxST_AgGberll4oxbCMeOFktM2-gjX37ZeRUeG7lDcwq71BEbm0gXfiCfLHVjFxA=w720-h310-rw', 'https://lh3.googleusercontent.com/VWBgjGMlcrJCywZhIXpCrmsS5GDiyx2yMRHETA8iQywIEyXE_x7zoLNjLIlt6QBBLuxD=w720-h310-rw', 'https://lh3.googleusercontent.com/I_bcuhPTp6cOEt1Oz23Rd6jSMVIrucMG3C8aoUURPiP5TL0dnjwXgmClTBB-2dCAXFQ=w720-h310-rw', 'https://lh3.googleusercontent.com/rNJSYE5IWmlK8Y8idAunJuqr0f1xYNgkLQdtXy9Ll_AOcPFOOTMYkRXrf_D8-dvKesQ=w720-h310-rw', 'https://lh3.googleusercontent.com/L6T_z74HEeZg0pl51ZpzBOftFuUDCDmIYEIvdfVhDv8lsG8_LbWDfBn16Jenuf04wg=w720-h310-rw']
video  :  None
category  :  ['GAME_MUSIC', 'FAMILY_MUSICVIDEO']
score  :  2.7
hist

## 3. Convert

In [16]:
# key represents initial panel, value represents panels (including the initial) tracking the initial panel
panels_track = {
'201812':['201812', '201912', '202001', '202003', '202004', '202009', '202010',
          '202011', '202012', '202101', '202102'],
'201907':['201907', '201908', '201909', '201912', '202001', '202003', '202004',
          '202009', '202010', '202011', '202012', '202101', '202102'],
'201908':['201908', '201912', '202001', '202003', '202004', '202009', '202010',
          '202011', '202012', '202101', '202102'],
'201912':['201912', '202001', '202003', '202004', '202009', '202010', '202011',
          '202012', '202101', '202102'],
'202001':['202001', '202003', '202004', '202009', '202010', '202011', '202012',
          '202101', '202102'],
'202002':['202002', '202004', '202009', '202010', '202011', '202012', '202101',
          '202102']}

for initial_panel, all_panels in panels_track.items():
    o = open_files(initial_panel=initial_panel,
                   all_panels=all_panels)
    app_details = o.open_initial_panel_with_its_tracking_panels()
    A = app_detail_dicts(d=app_details, all_panels=all_panels)
    merged_df = A.merge_panels_into_single_df()
    save = save_files(initial_panel=initial_panel, merged_df=merged_df)
    save.save_merged_df()
    print('panel data', initial_panel, 'has number of rows:')
    DF = merged_panel_df(merged_df=merged_df)
    DF.print_num_rows()

panel data 201812 has number of rows:
4278
panel data 201907 has number of rows:
10576
panel data 201908 has number of rows:
10537
panel data 201912 has number of rows:
7058
panel data 202001 has number of rows:
7027
panel data 202002 has number of rows:
6860


## 4. Descriptive Stats

In [3]:
# experiment cell for summary_statistics class
%run -i '2_summary_stats_visualization_new.py'

o = open_files(initial_panel='201907')
merged_df = o.open_merged_df()

DF1 = line_graphs(merged_df=merged_df,
                  all_panels=['201907', '201908', '201909', '201912',
                              '202001', '202003', '202004', '202009',
                              '202010', '202011', '202012', '202101', '202102'],
                  interval_start='201907',
                  interval_end='202102')
print(DF1.all_panels)
print(DF1.interval_start)

['201907', '201908', '201909', '201912', '202001', '202003', '202004', '202009', '202010', '202011', '202012', '202101', '202102']
201907


In [56]:
# experiment cell for missing imputation class
%run -i '2_summary_stats_visualization_new.py'

o = open_files(initial_panel='201907')
merged_df = o.open_merged_df()
all_panels = ['201907', '201908', '201909', '201912',
                              '202001', '202003', '202004', '202009',
                              '202010', '202011', '202012', '202101', '202102']
DF1 = impute_missing(merged_df=merged_df,
                  all_panels=all_panels,
                     missing_ratio = 0.3)


l1, l2 = DF1.cols_missing_ratio()
DF2 = DF1.drop_cols(list_of_col_names=l2)
# Now create a new class instance with DF2
new_class_df = impute_missing(merged_df=DF2,
                  all_panels=all_panels,
                     missing_ratio = 0.3)
A, l3 = new_class_df.check_apps_with_consecutive_missing_panels(var='minInstalls',
                                                         number_consec_panels_missing=2)
DF3 = new_class_df.drop_rows(list_of_row_labels=l3)
print(DF3.shape)
new_class_df = impute_missing(merged_df=DF3,
                  all_panels=all_panels,
                     missing_ratio = 0.3)
print(new_class_df.df.shape)

total number of columns contain missing value above 0.3 is 6
out of total number of columns 474
[('inAppProductPrice_202009', 0.4831694402420575), ('inAppProductPrice_202010', 0.48449319213313163), ('inAppProductPrice_202011', 0.4854387291981846), ('inAppProductPrice_202012', 0.48969364599092285), ('inAppProductPrice_202101', 0.4917738275340393), ('inAppProductPrice_202102', 0.4937594553706505)]
Before dropping we have 474 columns
After dropping we have 468 columns
['minInstalls_201907', 'minInstalls_201908', 'minInstalls_201909', 'minInstalls_201912', 'minInstalls_202001', 'minInstalls_202003', 'minInstalls_202004', 'minInstalls_202009', 'minInstalls_202010', 'minInstalls_202011', 'minInstalls_202012', 'minInstalls_202101', 'minInstalls_202102']
Before keeping we have 468 columns
After keeping we have 13 columns
number of apps with at least 2 consecutive missing panels for minInstalls are 339
out of 10576 apps.
Before dropping we have 10576 rows
After dropping we have 10237 rows
(1023

cols with too many missing values that you must drop:
inAppProductPrice_202009-202102 (the only columns that have missing ratio above 0.3)

rows (appids) with too many consecutively missing panels in important variables (etc minInstalls)
should be dropped. I found appids with at least 2 consecutively missing minInstalls are just 270 out of
10576 total apps, so just drop them.

sequence first drop columns then drop rows.

In [57]:
%run -i '2_summary_stats_visualization_new.py'
DF4 = new_class_df.impute_the_missing_panel_according_to_adjacent_panel(var='minInstalls')
DF5 = DF4[DF4.isnull().any(axis=1)]
print(DF5.shape)










['minInstalls_201907', 'minInstalls_201908', 'minInstalls_201909', 'minInstalls_201912', 'minInstalls_202001', 'minInstalls_202003', 'minInstalls_202004', 'minInstalls_202009', 'minInstalls_202010', 'minInstalls_202011', 'minInstalls_202012', 'minInstalls_202101', 'minInstalls_202102']
Before keeping we have 468 columns
After keeping we have 13 columns
(0, 13)
